cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginInstantiator.java

```diff
@@ -211,17 +208,32 @@ public ClassLoader getArtifactClassLoader(ArtifactId artifactId) throws IOExcept
     }
   }
 
-  private PluginProperties substituteMacroFields(Plugin plugin, MacroEvaluator macroEvaluator) {
+  private PluginProperties substituteMacros(Plugin plugin, @Nullable MacroEvaluator macroEvaluator) {
     Map<String, String> properties = new HashMap<>();
     Map<String, PluginPropertyField> pluginPropertyFieldMap = plugin.getPluginClass().getProperties();
-    MacroParser macroParser = new MacroParser(macroEvaluator);
+
+    // create macro evaluator and parser based on if it is config or runtime
+    boolean configTime = macroEvaluator == null;
+    TrackingMacroEvaluator trackingMacroEvaluator = new TrackingMacroEvaluator();
+    MacroParser macroParser;
+    if (configTime) {
+      macroParser = new MacroParser(trackingMacroEvaluator);
+    } else {
+      macroParser = new MacroParser(macroEvaluator);
+    }
+
     for (Map.Entry<String, PluginPropertyField> pluginEntry : pluginPropertyFieldMap.entrySet()) {
-      if (pluginEntry.getValue().isMacroSupported()) {
-        String macroValue = plugin.getProperties().getProperties().get(pluginEntry.getKey());
-        properties.put(pluginEntry.getKey(), macroParser.parse(macroValue));
-      } else {
-        properties.put(pluginEntry.getKey(), plugin.getProperties().getProperties().get(pluginEntry.getKey()));
+      String propertyValue = plugin.getProperties().getProperties().get(pluginEntry.getKey());
+      if (pluginEntry.getValue().isMacroSupported() && propertyValue != null) {
+        if (configTime) {
+          macroParser.parse(propertyValue);
```

Comment:
he is using parser to check syntax, the parser uses evaluator, which provides info if a field contained a macro or not after parsing is complete, a comment would be useful here to explain it.